cmd/snap-seccomp/main.go

```diff
@@ -789,28 +796,137 @@ func complainAction() seccomp.ScmpAction {
 	return seccomp.ActAllow
 }
 
+var osCreateTemp = os.CreateTemp
+
+func exportBPF(filter *seccomp.ScmpFilter) (io.ReadCloser, int64, error) {
+	errPrefixFmt := "cannot export bpf filter: %w"
+
+	filterFile, err := osCreateTemp("", "filter-file")
+	if err != nil {
+		return nil, 0, fmt.Errorf(errPrefixFmt, err)
+	}
+	if err := os.Remove(filterFile.Name()); err != nil {
+		return nil, 0, fmt.Errorf(errPrefixFmt, err)
+	}
+	// XXX: would be nice if ExportBPF would support an io.Writer
+	// but it requires a os.File
+	if err := filter.ExportBPF(filterFile); err != nil {
+		return nil, 0, fmt.Errorf(errPrefixFmt, err)
+	}
+	if _, err := filterFile.Seek(0, 0); err != nil {
+		return nil, 0, fmt.Errorf(errPrefixFmt, err)
+	}
+	stat, err := filterFile.Stat()
+	if err != nil {
+		return nil, 0, fmt.Errorf(errPrefixFmt, err)
+	}
+
+	return filterFile, stat.Size(), nil
+}
+
+// keep in sync with seccomp-support.c
+type scSeccompFileHeader struct {
+	header  [2]byte
+	version byte
+	// flags
+	unrestricted byte
+	// unused
+	padding [4]byte
+	// location of allow/deny, all offsets/len in bytes
+	lenAllowFilter uint32
+	lenDenyFilter  uint32
+	// reserved for future use
+	reserved2 [112]byte
+}
+
+func writeUnrestrictedFilter(out string) error {
+	hdr := scSeccompFileHeader{
+		header:  [2]byte{'S', 'C'},
+		version: 0x1,
+		// tell snap-confine
+		unrestricted: 0x1,
+	}
+	fout, err := osutil.NewAtomicFile(out, 0644, 0, osutil.NoChown, osutil.NoChown)
+	if err != nil {
+		return err
+	}
+	defer fout.Cancel()
+
+	if err := binary.Write(fout, arch.Endian(), hdr); err != nil {
+		return err
+	}
+	return fout.Commit()
+}
+
+func writeSeccompFilter(outFile string, filterAllow, filterDeny *seccomp.ScmpFilter) error {
+	allowBpf, allowSize, err := exportBPF(filterAllow)
+	if err != nil {
+		return err
+	}
+	defer allowBpf.Close()
+
+	denyBpf, denySize, err := exportBPF(filterDeny)
+	if err != nil {
+		return err
+	}
+	defer denyBpf.Close()
+
+	fout, err := osutil.NewAtomicFile(outFile, 0644, 0, osutil.NoChown, osutil.NoChown)
+	if err != nil {
+		return err
+	}
+	defer fout.Cancel()
+
+	hdr := scSeccompFileHeader{
+		header:  [2]byte{'S', 'C'},
+		version: 0x1,
+	}
+	hdr.lenAllowFilter = uint32(allowSize)
+	hdr.lenDenyFilter = uint32(denySize)
+	// silly go does not give an easy way to use native endian so just
+	// write big endian and use ntohl() in the C code
```

Comment:
This comment says `just write big endian`, but the `binary.Write` uses `arch.Endian()`?

Reply:
Nice catch! I indeed started with big endian but now that we have "arch.Endian()" this is no longer needed and it feels much more natural this way. I removed the comment :)